trio/_core/_thread_cache.py

```diff
@@ -68,9 +80,21 @@ def _work(self):
                 # 'deliver' triggers a new job, it can be assigned to us
                 # instead of spawning a new thread.
                 self._thread_cache._idle_workers[self] = None
-                deliver(result)
+                try:
+                    deliver(result)
+                except KillThisThread:
+                    return
+                except BaseException:
+                    # There's nothing else useful to do with it, and
+                    # if we let it escape from _work, there will be a
+                    # thread in _idle_workers that's not running but
+                    # can still have jobs assigned --> deadlock.
+                    logging.getLogger("trio.lowlevel.start_thread_soon").exception(
+                        f"Error delivering result {result!r} of work {fn!r}"
+                    )
                 del fn
                 del deliver
+                del result
```

Comment:
The lack of this `del` was keeping guest_state alive (via traceback locals) past the gc_collect_harder() in some of the internal error tests, and made the corresponding errors on destruction attach themselves to unrelated tests.